File: setupFigmaPages.js
// setupFigmaPages.js
// Purpose: Creates standard page structure in a Figma file using the REST API.
// Usage:
// 1. Ensure FIGMA_PAT environment variable is set: export FIGMA_PAT='your-token'
// 2. Run: node setupFigmaPages.js

const fetch = require('node-fetch');

// --- FILE_ID set to your Figma project ID ---
const FILE_ID = '368236963';

const FIGMA_PAT = process.env.FIGMA_PAT; // Reads PAT from environment variable

const pagesToCreate = [
  '_Foundations & Variables',
  '_Core Components [Adapted]',
  'Web - Dashboard',
  'Web - Agents',
  'Web - Personas',
  'Web - Memory',
  'Web - Projects',
  'Web - Settings',
  'Web - TrainingGround',
  'Android - Core Screens',
  'Prototypes',
  'Archive'
];

// --- Helper function to create a single page ---
async function createPage(fileId, pageName, accessToken) {
  const apiUrl = `https://api.figma.com/v1/files/${fileId}/pages`;
  console.log(`📄 Attempting to create page: ${pageName}...`);
  try {
    const response = await fetch(apiUrl, {
      method: 'POST',
      headers: {
        'X-Figma-Token': accessToken, // Correct header for PAT
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ name: pageName })
    });

    if (!response.ok) {
      // Log detailed error from Figma API if possible
      let errorDetails = '';
      try {
        const errorJson = await response.json();
        errorDetails = JSON.stringify(errorJson);
      } catch (e) {
        errorDetails = await response.text();
      }
      console.error(`   ❌ Failed: ${response.status} ${response.statusText}. Details: ${errorDetails}`);
      return false; // Indicate failure
    } else {
      // Success
      // const result = await response.json(); // Contains new page node_id if needed
      console.log(`   ✅ Success: Page '${pageName}' created.`);
      return true; // Indicate success
    }
  } catch (error) {
    console.error(`   ❌ Network or script error creating page '${pageName}':`, error);
    return false; // Indicate failure
  }
}

// --- Main execution function ---
async function main() {
  console.log("🚀 Starting Figma page setup script...");

  // Validate prerequisites
  if (!FIGMA_PAT) {
    console.error('❌ Error: FIGMA_PAT environment variable not set.');
    console.error('   Please set it using: export FIGMA_PAT="YOUR_TOKEN"');
    process.exit(1); // Exit if PAT is missing
  }

  console.log(`   Target Figma File ID: ${FILE_ID}`);
  console.log(`   Pages to create: ${pagesToCreate.length}`);

  let successCount = 0;
  let failureCount = 0;

  // Loop through pages and attempt creation
  for (const page of pagesToCreate) {
    const success = await createPage(FILE_ID, page, FIGMA_PAT);
    if (success) {
      successCount++;
    } else {
      failureCount++;
    }
    // Add a small delay to avoid hitting API rate limits aggressively
    await new Promise(resolve => setTimeout(resolve, 250)); // 250ms delay
  }

  console.log('🏁 Page setup script finished.');
  console.log(`   Summary: ${successCount} pages created successfully, ${failureCount} failed.`);

  if (failureCount > 0) {
      console.warn("⚠️ Please review errors above and potentially clean up/re-run if needed.");
  }
}

// --- Run the script ---
main().catch(error => {
  console.error("🚨 Unhandled error during script execution:", error);
  process.exit(1);
});

File: setup_figma_pages.py
#!/usr/bin/env python3
"""
setup_figma_pages.py - Creates standard page structure in a Figma file using the REST API.

Usage:
1. Run with PAT as an argument: python3 setup_figma_pages.py YOUR_FIGMA_PAT
   OR
   Set FIGMA_PAT environment variable: export FIGMA_PAT='your-token'
   Then run: python3 setup_figma_pages.py
"""

import os
import sys
import json
import time
import requests
from typing import List, Dict, Any, Optional, Tuple

# --- FILE_ID set to your Figma project ID ---
FILE_ID = '368236963'

# List of pages to create
PAGES_TO_CREATE = [
    '_Foundations & Variables',
    '_Core Components [Adapted]',
    'Web - Dashboard',
    'Web - Agents',
    'Web - Personas',
    'Web - Memory',
    'Web - Projects',
    'Web - Settings',
    'Web - TrainingGround',
    'Android - Core Screens',
    'Prototypes',
    'Archive'
]

def create_page(file_id: str, page_name: str, access_token: str) -> bool:
    """
    Create a single page in the Figma file.
    
    Args:
        file_id: Figma file ID
        page_name: Name of the page to create
        access_token: Figma Personal Access Token
        
    Returns:
        bool: True if successful, False otherwise
    """
    api_url = f"https://api.figma.com/v1/files/{file_id}/pages"
    
    print(f"📄 Attempting to create page: {page_name}...")
    
    headers = {
        'X-Figma-Token': access_token,
        'Content-Type': 'application/json'
    }
    
    data = {
        'name': page_name
    }
    
    try:
        response = requests.post(api_url, headers=headers, json=data)
        
        if not response.ok:
            # Try to get detailed error info
            try:
                error_details = response.json()
                error_str = json.dumps(error_details)
            except:
                error_str = response.text
                
            print(f"   ❌ Failed: {response.status_code} {response.reason}. Details: {error_str}")
            return False
        else:
            print(f"   ✅ Success: Page '{page_name}' created.")
            return True
    except Exception as e:
        print(f"   ❌ Network or script error creating page '{page_name}': {str(e)}")
        return False

def main():
    print("🚀 Starting Figma page setup script...")
    
    # Try to get PAT from command line args first, then environment variable
    figma_pat = None
    if len(sys.argv) > 1:
        figma_pat = sys.argv[1]
        print("   Using Figma PAT from command line argument.")
    else:
        figma_pat = os.environ.get('FIGMA_PAT')
        if figma_pat:
            print("   Using Figma PAT from environment variable.")
    
    # Validate prerequisites
    if not figma_pat:
        print("❌ Error: Figma PAT not provided.")
        print("   Please provide it as a command line argument:")
        print("   python3 setup_figma_pages.py YOUR_FIGMA_PAT")
        print("   OR set the FIGMA_PAT environment variable:")
        print("   export FIGMA_PAT='your-token'")
        sys.exit(1)
    
    print(f"   Target Figma File ID: {FILE_ID}")
    print(f"   Pages to create: {len(PAGES_TO_CREATE)}")
    
    success_count = 0
    failure_count = 0
    
    # Loop through pages and attempt creation
    for page in PAGES_TO_CREATE:
        success = create_page(FILE_ID, page, figma_pat)
        if success:
            success_count += 1
        else:
            failure_count += 1
        
        # Add a small delay to avoid hitting API rate limits
        time.sleep(0.25)  # 250ms delay
    
    print("🏁 Page setup script finished.")
    print(f"   Summary: {success_count} pages created successfully, {failure_count} failed.")
    
    if failure_count > 0:
        print("⚠️ Please review errors above and potentially clean up/re-run if needed.")
        return 1
    
    return 0

if __name__ == "__main__":
    try:
        exit_code = main()
        sys.exit(exit_code)
    except Exception as e:
        print(f"🚨 Unhandled error during script execution: {str(e)}")
        sys.exit(1)
